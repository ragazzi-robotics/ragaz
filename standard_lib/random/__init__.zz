from libc import srand, rand

var RAND_MAX: float = 0


@inline
def seed(s: uint):
    srand(s)

    # Try find out which the actual value of RAND_MAX
    # In GCC, RAND_MAX = 2147483647 = (2 ** 31) - 1, in other platforms, other number according to number of bits
    for bits in [16, 32, 64, 128]:
        var potential_randmax: float = ((2 ** (bits - 1)) - 1)
        var num: float = rand() as float / potential_randmax
        if num >= 0.0 and num <= 1.0:
            RAND_MAX = potential_randmax
            break


@inline
def random() -> float:
    return rand() / RAND_MAX


def randint(min: int, max: int) -> int:
    return (rand() % ((max + 1) - min)) + min


def uniform(min: float, max: float) -> float:
    return ((rand() / RAND_MAX) * (max - min)) + min
