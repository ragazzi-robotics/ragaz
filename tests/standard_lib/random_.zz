import random


def main():
    random.seed(100)

    for i in range(2):
        print(random.random())

    for i in range(6):
        print(random.randint(-50, 50))

    for i in range(6):
        print(random.uniform(-1, 1))
