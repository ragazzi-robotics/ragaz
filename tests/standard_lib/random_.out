0.240601
0.484127
10
36
-31
-28
12
-15
-0.183770
-0.699731
0.173103
0.263270
0.227719
-0.177322
